### PR TITLE
Feat: NpcTalkHandler / Fix: QuestHandler

### PR DESCRIPTION
**Pull request by angelotadeucci**
- Rewrite of `ScriptParser` / `ScriptMetadata`;
- Implemented `NpcTalkHandler.cs`.

### QuestHandler
- Add next quest when completing a quest;
- Quest item rewards

### CODE REVIEW

**Review by AlanMorel — APPROVED** [review]

**Comment by Zintixx** [comment]
We probably should be filtering out KR/CN/JP locales in scripts too

File: MapleServer2/PacketHandlers/Game/NpcTalkHandler.cs [hunk 1]
```diff
@@ -122,55 +142,22 @@ private static void HandleContinue(GameSession session, PacketReader packet)
             ScriptMetadata scriptMetadata = npcTalk.IsQuest ? ScriptMetadataStorage.GetQuestScriptMetadata(npcTalk.QuestId) : ScriptMetadataStorage.GetNpcScriptMetadata(npcTalk.Npc.Id);
             ResponseType responseType = npcTalk.IsQuest ? ResponseType.Quest : ResponseType.Dialog;
 
-            if (npcTalk.ScriptId != 0 && scriptMetadata.Options.First(x => x.Id == npcTalk.ScriptId).AmountContent <= npcTalk.ContentIndex && scriptMetadata.Options.First(x => x.Id == npcTalk.ScriptId).Goto.Count == 0)
+            if (npcTalk.ScriptId != 0
+                && scriptMetadata.Options.First(x => x.Id == npcTalk.ScriptId).AmountContent <= npcTalk.ContentIndex
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/NpcTalkHandler.cs:146]: can you cache `scriptMetadata.Options.First(x => x.Id == npcTalk.ScriptId)` so you don't need to query it twice?

File: MapleServer2/PacketHandlers/Game/NpcTalkHandler.cs [hunk 2]
```diff
@@ -248,7 +209,75 @@ private static void HandleBeauty(GameSession session)
...
+                }
+                else
+                {
+                    if (scriptMetadata.Options.First(x => x.Id == npcTalk.ScriptId).Goto.Count == 0)
+                    {
+                        return npcTalk.ScriptId;
+                    }
+                    else
+                    {
+                        return scriptMetadata.Options.First(x => x.Id == npcTalk.ScriptId).Goto[index];
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/NpcTalkHandler.cs:277]: I think you can cache `scriptMetadata.Options.First(x => x.Id > npcTalk.ScriptId)` here too

**Review by AlanMorel — CHANGES_REQUESTED** [review]

**Comment by angelotadeucci** [comment]
#88

**Review by AlanMorel — APPROVED** [review]

**Review by Sparkymod — APPROVED** [review]

File: MapleServer2/PacketHandlers/Game/NpcTalkHandler.cs [hunk 3]
```diff
@@ -217,6 +216,10 @@ private static DialogType GetDialogType(ScriptMetadata scriptMetadata, NpcTalk n
         {
             DialogType dialogType = DialogType.CloseNext1;
             if (scriptMetadata.Options.First(x => x.Id == npcTalk.ScriptId).Goto.Count == 0)
+            {
+                dialogType = DialogType.CloseNext1;
+            }
+            if (scriptMetadata.Options.First(x => x.Id == npcTalk.ScriptId).AmountContent > 1)
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/NpcTalkHandler.cs:227]: ```scriptMetadata.Options.First(x => x.Id == npcTalk.ScriptId)``` should probably be set to a variable so it doesn't need to be queried twice

**Review by AlanMorel — CHANGES_REQUESTED** [review]

**Review by AlanMorel — APPROVED** [review]

**Pull request merged**